### PR TITLE
EFCore Image model update and cleanup

### PR DESCRIPTION
**Pull request by ferferga**
Some of the entities defined in https://github.com/jellyfin/jellyfin/pull/2293 have changed, one of them being the images, which now include blurhashes.

**Changes**
- Removed DisplayPreferencesDto in MediaBrowserModel, in favor of the new one defined in Jellyfin.Api
- Renamed **Artwork** entity to **Image** (for clarity, as suggested by @dkanada https://github.com/jellyfin/jellyfin/pull/2293#issuecomment-619549080)
- Added different fields for tracking file creation/modification date and database creation/modification date.
- ImageType has been moved to ``Jellyfin.Data.Enums`` from ``MediaBrowser.Model.Entities``, to remove the duplication that previously existed with ``ArtworkKind``.
- Created a migration that updates the model as described below, being skipped if ``MigrateUserDb`` is not applied yet.
- Log a message when there are pending EFCore migrations, informing the user that they can take a while to complete and advising against shutting down the server while they're in progress.
- Merged **ImageInfos** with **Image**. Detail of the changes:

**New Image entity properties**

```
public int Id
public string Path
public ImageType Type
public string? Blurhash
public DateTime LastModifiedDate = DateTime.UtcNow;
public DateTime AddedDate
public DateTime? FileCreationDate
public DateTime? FileModificationDate
public uint RowVersion
```

*With ImageType being the current ImageType enum*

**Current Image entity properties**
```
public int Id
public string Path
public ArtKind Kind
public uint RowVersion
```

**Current ImageInfo properties**
```
public int Id
public Guid? UserId
public string Path
public DateTime LastModified
```

Every image we have in the client has (or can potentially have) all those fields in common. There's now a single 1:M or 1:1 relationship between any entity and ``Image``. Examples:

```
* Movie => 1 : M being M Image objects for the Primary, Background, Art, etc images
* Photo => 1 : 1 Photo entity storing shutter speed, GPS location, etc... And Image (type Primary, the default one) storing blurhash and the path on disk.
* User => 1 : 1 User storing all the user information and Image the profile photo.
```

In this PR, only used entities were changed to use the new ``Image`` entity, as the others haven't been migrated to EFCore yet and we have time to revisit the others as well.

**Issues**

~~After running the migration, ``Microsoft.EntityFrameworkCore.DbUpdateConcurrencyException`` exceptions are thrown after doing calls to the API controllers, not sure why (I call ``SaveChanges`` during the migration). **Stopping Jellyfin and starting it again fixes the problem and everything starts working correctly**.~~

Fixed on **Use EFCore migrations only to migrate to the new image model** commit.

### CODE REVIEW

**Comment by ferferga** [comment]
I did the model update in our own custom migration, instead of an EFCore model, not sure if for this kind of thing adding my custom code to the EFCore one was a better idea, I appreciate feedback on this :).

(Maybe this is the origin of the concurrency error)

File: Jellyfin.Data/Enums/ImageType.cs [hunk 1]
```diff
@@ -8,66 +8,66 @@ public enum ImageType
         /// <summary>
         /// The primary.
         /// </summary>
-        Primary = 0,
```
Comment (crobibero) [inline Jellyfin.Data/Enums/ImageType.cs:11]: Why did you remove all of the enum values?

Comment (ferferga) [reply Jellyfin.Data/Enums/ImageType.cs:11]: Afaik, enums are translated to int32 as in Java in C#, so I thought it was useless keeping them and thought it was a good change while I was at the cleanup process :)

However, I'm checking and all the enums in the codebase have their values specified, so maybe we want to have this reverted for consistency. Or apply this change throughout the code base (not for this PR definitely, but I can make it)

Thoughts? @jellyfin/server

**Comment by dkanada** [comment]
I would suggest something like this to match the file columns. You also listed the FileModificationDate as non-public in the comment above but it looks to be public in the code itself which is good.

```
LastModifiedDate -> EntityModificationDate
AddedDate -> EntityInsertionDate
```

**Comment by ferferga** [comment]
@dkanada I'm not able to find the terminology you mention in the codebase anywhere, can you point me an example?

Imo the current terminology is clear enough to be distinguished between the file one and the entity one while being short. ``AddedDate`` to ``InsertionDate`` is a good change though.

File: Jellyfin.Data/Entities/Image.cs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+        /// <summary>
+        /// Gets the creation date of the file.
+        /// </summary>
+        public DateTime? FileCreationDate { get; private set; }
+
+        /// <summary>
+        /// Gets the modification date of the file.
+        /// </summary>
+        public DateTime? FileModificationDate { get; private set; }
```
Comment (dkanada) [inline Jellyfin.Data/Entities/Image.cs:93]: I just meant these items are named somewhat inconsistently. One option would be the one I suggested yesterday, and another one might be combining the two modified properties (they seem a bit redundant) and removing the `File` prefix completely.

Comment (ferferga) [reply Jellyfin.Data/Entities/Image.cs:93]: Right now ofc they're redundant but I can see when this can be useful once everything uses this entity. Examples:

* **Change the type of an image**: This is currently impossible with the API, but we can have this in the future, and show the date of that change in the UI. The change of the type of an image doesn't imply that Blurhash changes, so the computation can be skipped

* **File modified on disk**: We could have modified the properties of the image on the UI, but also some people might edit the library images in place (or use a compression script to reduce their sizes), which means that we need to recompute their blurhashes, which is something we can do if we detect a mismatch during scanning time and the database value. Also, if the scanner gets interrupted when this situation happen, the scanner will be able to catch up as it will detect that the database last modified date is older than the file's.

The added dates I can agree with you that are less useful. I just put them there so we can expose that information in the API the same way we do with media items.

**Comment by MrTimscampi** [comment]
As was discussed in the team channel, I'd be nice to have two IDs for stuff in the database:

* Integers as the Primary Key. Generally more efficient than using UUIDs/GUIDs when doing comparisons (Depends on the database used, but if we aim to support multiple DB softwares, this is the most efficient thing to do)
* UUIDs (Valid ones, not GUIDs) as "external keys" for API usage.

This gives us the performance from working with integers internally, while preventing issues that arise when using sequential IDs in an API.

**Comment by ferferga** [comment]
Another thing commented is the possibility of having a ``File`` entity that is joined with the ``Image`` one, with fields like the File Modification and File Creation Date, plus maybe MD5. 

Any entity (like movies, episodes...) that are linked to a file will reference the ID of that File table.

File: Jellyfin.Server/CoreAppHost.cs [hunk 3]
```diff
@@ -74,7 +74,8 @@ protected override void RegisterServices()
             }
 
             ServiceCollection.AddDbContextPool<JellyfinDb>(
-                 options => options.UseSqlite($"Filename={Path.Combine(ApplicationPaths.DataPath, "jellyfin.db")}"));
+                 options => options.UseLazyLoadingProxies().
```
Comment (cvium) [inline Jellyfin.Server/CoreAppHost.cs:77]: Why?

Comment (ferferga) [reply Jellyfin.Server/CoreAppHost.cs:77]: @cvium I've been stuck for hours trying to check why EFCore wouldn't load the ``Image`` entity related to an user, thinking something was borked in my migration procedures. Rest of the code worked fine (i.e I could login and see my profile image) but somehow I couldn't access ``user.ProfileImage`` (it was null) in a ``Jellyfin.Server`` migration I did for testing that everything was accessible.

Researching about this, I found that EFCore doesn't load related entities by default and that you must explicitly include them (https://docs.microsoft.com/es-es/ef/core/querying/related-data/eager), otherwise they're ``null`` if they're nullable in the code.

Adding ``virtual`` to the properties allows EFCore to override the getter and do the query for you without further work (which I think was the initial design goal of the schema as everything was ``virtual``). However, this needs of LazyLoading (https://docs.microsoft.com/es-es/ef/core/querying/related-data/lazy) and ``Microsoft.EntityFrameworkCore.Proxies`` dependency to work.

So basically I added this so everything is easily accessible out of the box and nobody else spends more time debugging this the same way I did just because I wasn't properly aware of this EFCore feature. I don't mind not reverting the commit as I know now how to do proper queries with ``nullable`` stuff, but I think it's nice to have.

Here you have all the ways of loading data using EFCore: https://docs.microsoft.com/es-es/ef/core/querying/related-data/

Comment (cvium) [reply Jellyfin.Server/CoreAppHost.cs:77]: I don't like lazy loading tbh. It creates separate queries for every relationship when accessed and it doesn't work if you disconnect the entity from the db context. If some unfortunate soul decides to map all properties to a Dto, the lazy loading would bite them in the ass.

Comment (ferferga) [reply Jellyfin.Server/CoreAppHost.cs]: @cvium Fine, reverted! 👍

File: Jellyfin.Data/Enums/ImageType.cs (hunk 1 above)
Comment (EraYaN) [reply Jellyfin.Data/Enums/ImageType.cs:11]: Specified values is better IMO

File: Jellyfin.Server.Implementations/JellyfinDbProvider.cs [hunk 4]
```diff
@@ -13,19 +15,27 @@ public class JellyfinDbProvider
...
         {
             _serviceProvider = serviceProvider;
             _appPaths = appPaths;
+            _logger = logger;
 
             using var jellyfinDb = CreateContext();
-            jellyfinDb.Database.Migrate();
+            if (jellyfinDb.Database.GetPendingMigrations().Any())
+            {
+                _logger.LogInformation("There are pending EFCore migrations in the database. Applying... (This may take a while, do not stop Jellyfin)");
```
Comment (EraYaN) [inline Jellyfin.Server.Implementations/JellyfinDbProvider.cs:35]: EF Core should log this itself as well, so we might just open up that SourceContext to the output?

**Comment by EraYaN** [comment]
On the potential `File` entity.

the https://github.com/jellyfin/jellyfin/blob/master/Jellyfin.Data/Entities/Libraries/MediaFile.cs one is essentially that. And it has https://github.com/jellyfin/jellyfin/blob/master/Jellyfin.Data/Entities/Libraries/MediaFileStream.cs children.

This PR has the diagram: https://github.com/jellyfin/jellyfin/pull/2293

diagram_2020-04-26_001

The `Release` type is the metadata -> physical set of files connection. So that also hooks up all the sidecars (subtitles, external audio, XMP files, editing decision lists etc). Then the `MediaFile` entity to point to actual file locations.

The original `Artwork` entity was always meant to contain only metadata images and other system images. Photo libraries should use the Media* classes so all libraries work the same internally.

File: Jellyfin.Data/Enums/ImageType.cs (hunk 1 above)
Comment (ferferga) [reply Jellyfin.Data/Enums/ImageType.cs]: Makes sense, hadn't thought about it @barronpm 

Adressed!

File: Jellyfin.Server.Implementations/JellyfinDbProvider.cs (hunk 4 above)
Comment (ferferga) [reply Jellyfin.Server.Implementations/JellyfinDbProvider.cs:35]: Added in [CoreAppHost](https://github.com/jellyfin/jellyfin/pull/6049/files#diff-4d188b93a5a2169482c7c60b65a4fd098d2131c2b8c48622abdfb80efe125138R77)

File: Jellyfin.Server/Program.cs [hunk 5]
```diff
@@ -224,7 +224,7 @@ private static async Task StartApp(StartupOptions options)
             {
                 _logger.LogInformation("Running query planner optimizations in the database... This might take a while");
                 // Run before disposing the application
-                using var context = new JellyfinDbProvider(appHost.ServiceProvider, appPaths).CreateContext();
+                using var context = new JellyfinDbProvider(appHost.ServiceProvider, appPaths, (ILogger<JellyfinDbProvider>)_logger).CreateContext();
```
Comment (cvium) [inline Jellyfin.Server/Program.cs:227]: JellyfinDbProvider is already registered in ServiceProvider, so you could just call `appHost.Resolve<JellyfinDbProvider>()`

File: Jellyfin.Server.Implementations/Migrations/20210518000710_NewImageModel.cs [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+                    table: "ImageInfos",
+                    type: "TEXT",
+                    nullable: true,
+                    defaultValue: new Guid("00000000-0000-0000-0000-000000000000"));
+
+                var triggerQuery = @"CREATE TRIGGER AutoGenerateGUID
+                                    AFTER UPDATE ON ImageInfos
+                                    FOR EACH ROW
+                                    BEGIN
+                                        UPDATE ImageInfos SET UUID = (select hex(randomblob(4)) || '-' || hex(randomblob(2))
```
Comment (cvium) [inline Jellyfin.Server.Implementations/Migrations/20210518000710_NewImageModel.cs:129]: How is the performance on this exactly? Can't say I'm a fan of triggers

Comment (ferferga) [reply Jellyfin.Server.Implementations/Migrations/20210518000710_NewImageModel.cs:129]: @cvium Both this migration and the ``AddIndexesAndCollations`` one (the only two pending migrations since 10.7.X) run in less than 0.5 seconds in my Raspberry Pi. This trigger is just a "workaround" around the fact that we can't do ``SELECT``s in migrationBuilder and SQLite doesn't support functions, so I can't do a programatic update that allows me to generate a GUID for each row in the table.

The trigger is the best "programatic" method SQLite has for the kind of behaviour we have, and it's dropped as soon as all the migration procedure finishes, so it won't affect performance outside the migration process itself. New installations will also have no performance hit, as the tables will be empty.

Comment (cvium) [reply Jellyfin.Server.Implementations/Migrations/20210518000710_NewImageModel.cs:129]: I missed the line where you drop the trigger. What makes you say we can't do SELECTs in migrationbuilder though?

Comment (ferferga) [reply Jellyfin.Server.Implementations/Migrations/20210518000710_NewImageModel.cs:129]: @cvium We can do SELECTs in subqueries (in fact I do them in many of the other statements as subqueries), however we can't get the returned values in C#, so we can't do a ``foreach (var row in query)`` like we do in our custom migration runner, as migrationBuilder returns an instance of [``operationBuilder``](https://docs.microsoft.com/es-es/dotnet/api/microsoft.entityframeworkcore.migrations.operations.builders.operationbuilder-1?view=efcore-5.0).

With raw SQL, I'm not able to do an UPDATE operation where the SELECT that generates UUIDs generates an unique UUID for each row. The only way of doing it is using a trigger, which runs an statement per updated row.

**Comment by ferferga** [comment]
There's an upstream bug causing the entire row to be deleted when setting related entities to null in EFCore that blocks this PR. https://github.com/dotnet/efcore/issues/22196

Labeling as blocked until EFCore 6 is in or the backport to EFCore 5 is done (hopefully)

**Comment by barronpm** [comment]
> There's an upstream bug causing the entire row to be deleted when setting related entities to null in EFCore that blocks this PR. [dotnet/efcore#22196](https://github.com/dotnet/efcore/issues/22196)
> 
> Labeling as blocked until EFCore 6 is in or the backport to EFCore 5 is done (hopefully)

The linked bug is specifically for composite foreign keys, but that doesn't seem relevant here

**Comment by ferferga** [comment]
@barronpm However, [this comment](https://github.com/dotnet/efcore/issues/22196#issuecomment-689652731) in the thread seems to be the exact same situation we're facing: ``ProfileImage`` can be null but ``Id`` in ``Image`` isn't.

**Comment by ferferga** [comment]
@jellyfin-bot rebase

**Comment by ferferga** [comment]
Cherry-picked some commits into #6522

File: Jellyfin.Server.Implementations/JellyfinDbProvider.cs [hunk 7]
```diff
@@ -13,19 +15,27 @@ public class JellyfinDbProvider
...
+        /// <param name="logger">The logger.</param>
+        public JellyfinDbProvider(IServiceProvider serviceProvider, IApplicationPaths appPaths, ILogger<JellyfinDbProvider> logger)
         {
             _serviceProvider = serviceProvider;
             _appPaths = appPaths;
+            _logger = logger;
 
             using var jellyfinDb = CreateContext();
-            jellyfinDb.Database.Migrate();
+            if (jellyfinDb.Database.GetPendingMigrations().Any())
```
Comment (JPVenson) [inline Jellyfin.Server.Implementations/JellyfinDbProvider.cs:33]: migrations are already executed in ApplicationHost.cs L579

File: Jellyfin.Server.Implementations/Migrations/20210906152818_NewImageModel.cs [hunk 8]
```diff
@@ -0,0 +1,217 @@
+#pragma warning disable CS1591
+// <auto-generated />
+using System;
+using Jellyfin.Data.Enums;
+using Microsoft.EntityFrameworkCore.Migrations;
+
+namespace Jellyfin.Server.Implementations.Migrations
+{
+    public partial class NewImageModel : Migration
```
Comment (JPVenson) [inline Jellyfin.Server.Implementations/Migrations/20210906152818_NewImageModel.cs:9]: migrations now live in their respective providers assemblies

File: Jellyfin.Server.Implementations/Migrations/20210906152818_NewImageModel.cs [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+
+                // End of the custom migration procedures
+            }
+            else
+            {
+                migrationBuilder.DropTable(
+                    name: "ImageInfos",
+                    schema: "jellyfin");
+
+                migrationBuilder.CreateTable(
```
Comment (JPVenson) [inline Jellyfin.Server.Implementations/Migrations/20210906152818_NewImageModel.cs:166]: The library.db migration also introduced a ImageInfo table
https://github.com/jellyfin/jellyfin/blob/master/src/Jellyfin.Database/Jellyfin.Database.Implementations/Entities/BaseItemImageInfo.cs#L10
Could you repurpose that instead?

**Review by JPVenson — CHANGES_REQUESTED** [review]

**Comment by Bond-009** [comment]
Closing as stale.
